### PR TITLE
Added an argument mode manager and a demonstration of how it might be used in HaplotypeCaller --dragen-mode

### PR DESCRIPTION
**Pull request by jamesemery**

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/cmdline/ModeArgumentUtils.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+ * The detection of the mode (i.e. the implementation of the mode argument) is the tool's responsibility. Following
+ * that, a public method of this class, setArgValues, can be used to install the associated argument values
+ * in the command parser, which will propagate them into the associate tools instance fields.
+ *
+ * The setting of the arguments is done such that argument already specified on the command line are not
+ * overwritten. This allows for mode refinement - i.e. the setting of the mode while turing some of its arguments
+ * to specific values.
+ */
+
+public class ModeArgumentUtils {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/cmdline/ModeArgumentUtils.java:31]: `final`

File: src/main/java/org/broadinstitute/hellbender/cmdline/ModeArgumentUtils.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            if ( !hasBeenSet(parser, argValues[i]) ) {
+                String parserMessage = setValue(parser, argValues[i], argValues[i+1]);
+
+                if ( StringUtils.isEmpty(parserMessage) ) {
+                    modifiedArgs.put(argValues[i], argValues[i + 1]);
+                } else {
+                    modifiedArgs.put(argValues[i], argValues[i + 1] + " (" + parserMessage + ")");
+                }
+            } else {
+                logger.info("parameter not set by this mode, as it was already set on the command line: " + argValues[i]);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/cmdline/ModeArgumentUtils.java:61]: Include the mode name in the logger message

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCaller.java [hunk 3]
```diff
@@ -160,14 +161,18 @@ public List<ReadFilter> getDefaultReadFilters() {
...
     protected String[] customCommandLineValidation() {
         if (hcArgs.dragenMode) {
             final GATKReadFilterPluginDescriptor readFilterPlugin =
                     getCommandLineParser().getPluginDescriptor(GATKReadFilterPluginDescriptor.class);
             Optional<ReadFilter> filterOptional = readFilterPlugin.getResolvedInstances().stream().filter(rf -> rf instanceof MappingQualityReadFilter).findFirst();
             filterOptional.ifPresent(readFilter -> ((MappingQualityReadFilter) readFilter).minMappingQualityScore = 1);
+            ModeArgumentUtils.setArgValues(
+                    getCommandLineParser(),
+                    hcArgs.getDragenNameValuePairs(),
+                    HaplotypeCallerArgumentCollection.DRAGEN_GATK_MODE_LONG_NAME);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCaller.java:175]: Would it be a good idea for `setArgValues()` itself to check whether the provided mode is activated, and proceed with setting the args only if it's active? Or is an external check on the mode like you have here preferable? Either way, we don't need to address in this PR.

File: src/main/java/org/broadinstitute/hellbender/cmdline/ModeArgumentUtils.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+public class ModeArgumentUtils {
+
+    protected static final Logger logger = LogManager.getLogger(ModeArgumentUtils.class);
+
+    /**
+     * set a group of  arguments (and associated fields) to the specified values - thereby implementing a mode
+     *
+     * The method does not overwrite arguments already set on the command line (as indicated by the parser)
+     * Additionally, the method emits a warning message indicating which argument were set by the mode and
+     * to which values.
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/cmdline/ModeArgumentUtils.java:40]: done

File: src/main/java/org/broadinstitute/hellbender/cmdline/ModeArgumentUtils.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+     *
+     * @param parser - commandline parser associated with the tool. This should be an instance of CommandLineArgumentParser
+     * @param argValues - an array of arguments and values to potentially set. The order of elements in the array
+     *                  is arg0,value0,arg1,value1 ...
+     * @param modeName - the name of the mode being set. This is used in textual message, such as the warning
+     *                 issued to notify the user of the changed argument
+     */
+    public static void setArgValues(final CommandLineParser parser, final String[] argValues, final String modeName) {
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/cmdline/ModeArgumentUtils.java:48]: This method should only be called if the tool has already determined that it must be set. I added a note about it.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerArgumentCollection.java [hunk 6]
```diff
@@ -203,4 +210,23 @@ protected ReadThreadingAssemblerArgumentCollection getReadThreadingAssemblerArgu
...
+                ALLELE_EXTENSION_LONG_NAME, "1",
+                LikelihoodEngineArgumentCollection.DISABLE_CAP_BASE_QUALITIES_TO_MAP_QUALITY_LONG_NAME, "true",
+                LikelihoodEngineArgumentCollection.ENABLE_DYNAMIC_READ_DISQUALIFICATION_FOR_GENOTYPING_LONG_NAME, "true",
+                LikelihoodEngineArgumentCollection.EXPECTED_MISMATCH_RATE_FOR_READ_DISQUALIFICATION_LONG_NAME, "0.03",
+                GenotypeCalculationArgumentCollection.GENOTYPE_ASSIGNMENT_METHOD_LONG_NAME, String.valueOf(GenotypeAssignmentMethod.USE_POSTERIOR_PROBABILITIES),
+                AssemblyRegionArgumentCollection.INDEL_PADDING_LONG_NAME, "150",
+                GenotypeCalculationArgumentCollection.CALL_CONFIDENCE_LONG_NAME, "3.0",
+                GenotypeCalculationArgumentCollection.USE_POSTERIORS_TO_CALCULATE_QUAL_LONG_NAME, "true",
+        };
+    }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerArgumentCollection.java:231]: What about `hcArgs.informativeReadOverlapMargin = 1;` from the old settings?

Also, were `ReadFilterArgumentDefinitions.MINIMUM_MAPPING_QUALITY_NAME, "1"` and `ALLELE_EXTENSION_LONG_NAME, "1"` set using the old method? I don't see them

**Review by droazen — CHANGES_REQUESTED** [review]
@jamesemery Back to you with my comments

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerArgumentCollection.java (hunk 6 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerArgumentCollection.java:231]: `hcArgs.informativeReadOverlapMargin = 1;` is handled. That is the `ALLELE_EXTENSION_LONG_NAME, "1"` in that list. Its argument name and its argument collection variable are fairly confusingly discordant. 

`MAPPING_QUALITY_THRESHOLD_FOR_GENOTYPING_LONG_NAME, "1",` is also handled (which is distinct from the mapping quality on the read filter but that is handled  by `HaplotypeCaller.customCommandLineValidaiton()`

**Review by droazen — APPROVED** [review]
:+1: Merge when ready

**Pull request merged**